feat: send email notification to patient when doctor confirms or rejects appointment

nt-lan-anh opened the pull request:
## Related Tickets

- [TicketID](https://edu-redmine.sun-asterisk.vn/issues/92070)
- [TicketID](https://edu-redmine.sun-asterisk.vn/issues/92071)

## WHAT

Gửi email notification cho **bệnh nhân** khi bác sĩ **xác nhận** hoặc **từ chối** lịch hẹn.

## HOW

* `AppointmentServiceImpl`

  * Gọi `publishAppointmentConfirmedEvent(saved, slot)` ngay sau khi confirm & sync calendar.
  * Gọi `publishAppointmentRejectedEvent(saved, slot, request.reason())` ngay sau khi reject & sync calendar.
* `NotificationEventListener`

  * Đã lắng nghe thêm `AppointmentConfirmedEvent` và `AppointmentRejectedEvent` để chuyển tiếp sang `NotificationService`.
* `NotificationServiceImpl.buildRegistry()`

  * Thêm 2 mapping:

    * **Confirmed** → subject key: `email.subject.appointment.confirmed`, template: `email/appointment-confirmed`, gửi tới `patientEmail()`.
    * **Rejected** → subject key: `email.subject.appointment.rejected`, template: `email/appointment-rejected`, gửi tới `patientEmail()` (kèm `reason`).
  * Mapping sử dụng accessor từ record event: `appointmentId()`, `patientEmail()`, `patientName()`, `doctorName()`, `startTime()`, `endTime()`, `reason()` (với reject). &#x20;
* Templates (Thymeleaf)

## Evidence
[VIDEO DEMO](https://drive.google.com/file/d/141vKDLx40z63hgseCPhdSot2LUGu_du2/view?usp=drive_link)

nt-lan-anh commented:
Ready

vanvtt-0952 commented on backend/src/main/java/com/example/backend/service/impl/NotificationServiceImpl.java:
```diff
@@ -121,6 +123,40 @@ private Map<Class<?>, HandlerConfig> buildRegistry() {
+                    vars.put("startTime", ((AppointmentConfirmedEvent) e).startTime());
+                    vars.put("endTime", ((AppointmentConfirmedEvent) e).endTime());
+                    return vars;
+                }));
```
- Tách hàm dựng vars: Tạo hàm riêng để xây dựng biến thay vì để lambda phình to, rối mắt và logic
- Map.of() để tạo map bất biến vì chị thấy ko cần chỉnh sửa map sau này
- Khai báo Record cho HandlerConfig để giảm boilerplate (setter/getter)


tự xử cho method Patient nhận noti nhé
```
// Nếu HandlerConfig là bất biến, dùng record
public record HandlerConfig(
    NotificationType notificationType,
    String notificationTemplate,
    String emailSubject,
    String emailTemplate,
    Function<Object, Object[]> argsExtractor,
    Function<Object, String> emailExtractor,
    Function<Object, Map<String, Object>> varsExtractor
) {}

// Hàm dựng vars cho AppointmentConfirmedEvent
private Map<String, Object> buildAppointmentConfirmedVars(AppointmentConfirmedEvent event) {
    return Map.of(
        "appointmentId", event.appointmentId(),
        "patientName", event.patientName(),
        "doctorName", event.doctorName(),
        "startTime", event.startTime(),
        "endTime", event.endTime()
    );
}

// Thêm vào map với method reference và pattern matching
map.put(AppointmentConfirmedEvent.class,
    new HandlerConfig(
        NotificationType.APPOINTMENT_CONFIRMED,
        "notification.appointment.confirmed",
        "email.subject.appointment.confirmed",
        "email/appointment-confirmed",
        e -> new Object[]{((AppointmentConfirmedEvent) e).appointmentId()},
        e -> ((AppointmentConfirmedEvent) e).patientEmail(),
        e -> {
            if (e instanceof AppointmentConfirmedEvent ace) {
                return buildAppointmentConfirmedVars(ace);
            }
            return Map.of();
        }
    )
);
```

ntla-sunas committed:
feat: send email notification to patient when doctor confirms or rejects appointment

feat: send email notification to patient when doctor confirms or rejects appointment
